### ISSUE TITLE
URL encode / decode for uid

### ISSUE THREAD

**Issue by tftm** [devture/matrix-corporal#22]
Hi. We are trying to use the UserPrincipalName entry from MS AD as the UID field. In this case Synapse replaces the @ in the username with a slash (/) and performs the url_encode operation when accessing the API. Thus, the string /%40name%2Fdomain%3Adomain is formed in the URL.
However, matrix_corporal does not do this, and when we using name/domain format names, it substitutes them directly into the URL. As a result, for "id": "@name/domain:domain" record in policy.json the API is accessed at /_synapse/admin/v1
/users/@name/domain:domain/login which results in the error code=400 wrapped=M_UNRECOGNIZED: Unrecognized request

**Comment by spantaleev** [comment]
This should be fixed in `2.3.1`.

---

Would Synapse like you throwing things like `@name/domain:domain` there anyway?

**Comment by tftm** [comment]
We have many numeric accounts in AD, but Synapse has the specifics of working with them https://github.com/matrix-org/matrix-synapse-ldap3/issues/55
Using UserPrincipalName as uid is the only alternative I see.

**Comment by tftm** [comment]
@spantaleev there seems to be two encodings going on
In logs I see records like uri='/_synapse/admin/v1/users/%2540user%252Fdomain%253Adomain/login' 
but it is url_encode() of uri='/_synapse/admin/v1/users/%40user%2Fdomain%3Adomain/login'

**Comment by spantaleev** [comment]
You're right.. `2.3.1` made things worse - completely breaking reconcilliation :man_facepalming: 

I've since reverted this URL-encoding change. `2.3.2` should be pretty much the same as `2.3.0`.

People with servers with normal mxids should no longer see errors during reconcilliation.

----

Perhaps your problem is not with matrix-corporal's reconcilliation though? But with you hitting `/_synapse/admin/v1/users/<user_id>/login` yourself from your own scripts?

Or is reconcilliation failing on `2.3.0` (and now on `2.3.2`)?

**Comment by tftm** [comment]
Reconcilliation fails on 2.3.0 without creating first user in list.
On 2.3.1 the reconcilliation fails too, but the first user is created.
M.b. I'm wrong, but I guess, that on 2.3.1 matrix-corporal sends the correct uid first (encoded it on POST /login), than get response with same encoded uid as confirmation. 
For the second time matrix-corporal takes the already encoded uid and encoded it once more on POST /login.

M.b. something like this can solve the issue:
```
if strings.Contains(userId, "/") {
               userId  = url.QueryEscape(userId)
}
err = client.MakeRequest(
 		"POST",
 		buildPrefixlessURL(client, fmt.Sprintf("/_synapse/admin/v1/users/%s/login", userId), map[string]string{}),
 		requestPayload,
 		&response,
 	)
```
but I wouldn't break master once more)

**Comment by spantaleev** [comment]
You can try that patch yourself and see if it helps with all of your users.

Something is definitely fishy with these users containing `/`. I suspect it may be an upstream problem too. Perhaps the `/_synapse/admin/v1/users/<user_id>/login` API needs to be adjusted so that it handles URL-encoded values or something. It seems to hate URL-encoding in the general case, but like it in case there are things like `/`. Hmm..

Have you tried calling these APIs manually (with an admin token) to see what values they like and don't like? You could then report the issue to the Synapse repository.
